src/systemctl/systemctl-list-unit-files.c

```diff
@@ -11,25 +12,37 @@
 
 static int compare_unit_file_list(const UnitFileList *a, const UnitFileList *b) {
         const char *d1, *d2;
+        _cleanup_free_ char *a_id = NULL, *b_id = NULL;
+        int r;
 
         d1 = strrchr(a->path, '.');
         d2 = strrchr(b->path, '.');
 
         if (d1 && d2) {
-                int r;
-
                 r = strcasecmp(d1, d2);
                 if (r != 0)
                         return r;
         }
 
-        return strcasecmp(basename(a->path), basename(b->path));
+        r = path_extract_filename(a->path, &a_id);
+        if (r < 0)
+                return r;
+
+        r = path_extract_filename(b->path, &b_id);
+        if (r < 0)
+                return r;
```

Comment:
no, that's not ok. compare_unit_file_list() is for ordering two items, i.e. it#s a function for use in qsort(), basically.

Functions like that are not allowed to fail. Which kinda sucks. But given that this is how it is, maybe handle it as good as we possibly could, i.e. maybe compare the full paths in that case.

or in other words, just eat up the errors. like this:

```c
/* qsort() callbacks are not allowed to fail, hence handle failures gracefully, yet systematically by comparing the full paths insteads. */
(void) path_extract_filename(a->path, &a_id);
(void) path_extract_filename(b->path, &b_id);
return strcasecmp(a_id ?: a->path, b_id ?: b->path);
```

or something like that.

Reply:
or actually, much better than ths, just use the underlying path_find_last_component() helper here. it avoids memory allocation. hence it can only fail due to bad inputs, and in that case compare error codes.

Reply:
Would it be safe to do that?  `path_find_last_component` doesn't modify its string, so while "abc/./" and "abc/./././" are logically the same path, and would compare identically with `path_extract_filename`, they would differ using `path_find_last_component`.  (then again, this is meant to be files, not directories, so one hopes that wouldn't matter… not sure how "defensive" this code is meant to be)

Reply:
use memcmp_nn() on the pointer and the length path_find_last_component() return.

i.e.

```
int u = path_find_last_component(a->path, /* accept_dot_dot= */ false, /* next= */ NULL, &a_id);
int v = path_find_last_component(b->path, /* accept_dot_dot= */ false, /* next= */ NULL, &b_id);

if (u < 0 || v < 0) /* One failed? Then compare error codes */
        return CMP(u, v);

return ascii_strcasecmp_nn(a->id, u, b->id, v);
```